### PR TITLE
fix(runJob): cancel underlying job on cancel stage

### PR DESCRIPTION
**Pull request by emjburns**
Currently, when you cancel a run job stage it just leaves the underlying job running. That's unexpected behavior. 

This PR adds support for canceling the underlying job when you cancel the stage.

@ezimanyi @ethanfrogers I'm looking for some help verifying that this data structure also works for kubernetes (specifically that `jobStatus.name` is the name of the kubernetes job and the same name used in this [destroyJob](https://github.com/spinnaker/clouddriver/blob/29c5c1435ab0c750c4b491208f600647a3e98901/clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/v1/deploy/ops/job/DestroyKubernetesJobAtomicOperation.groovy) atomic operation).

This PR uses https://github.com/spinnaker/clouddriver/pull/3762 for Titus support.

### CODE REVIEW

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/job/RunJobStage.java [hunk 1]
```diff
@@ -16,20 +16,40 @@
...
 
 @Component
-public class RunJobStage implements StageDefinitionBuilder {
+public class RunJobStage implements StageDefinitionBuilder, CancellableStage {
+
+  private final Logger log = LoggerFactory.getLogger(getClass());
+  private final DestroyJobTask destroyJobTask;
+  private final ObjectMapper objectMapper = OrcaObjectMapper.newInstance();
+
+  public RunJobStage() {
```
Comment (cfieber) [inline orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/job/RunJobStage.java]: why is this needed? I suspect it is just so existing tests don't need to be updated... I'd lean towards just updating the tests to pass in a stub if that is the case

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/job/RunJobStage.java [hunk 2]
```diff
@@ -56,6 +76,43 @@ public void taskGraph(Stage stage, TaskNode.Builder builder) {
...
+      destroyContext.put("cloudProvider", context.getCloudProvider());
+      destroyContext.put("region", context.getJobStatus().getRegion());
+      destroyContext.put("credentials", context.getCredentials());
+
+      stage.setContext(destroyContext);
+
+      destroyJobTask.execute(stage);
+    } catch (Exception e) {
+      log.error(
+          String.format(
```
Comment (marchello2000) [inline orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/job/RunJobStage.java]: nit: some weird parens here. also, for my education, is `RunJob` titus specific? if not, probably worth making this log message more generic

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/job/RunJobStage.java (hunk 1 above)
Comment (emjburns) [reply orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/job/RunJobStage.java]: You are correct, I'll fix the tests so this is not needed.

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/job/RunJobStage.java (hunk 2 above)
Comment (emjburns) [reply orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/job/RunJobStage.java]: Oops good call. Run job is for titus and kubernetes, so I'll make this generic. I'll also fix the parens

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/job/RunJobStageContext.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+  public Map<String, Object> other() {
+    return other;
+  }
+
+  @JsonAnySetter
+  public void set(String name, Object value) {
+    other.put(name, value);
+  }
+
+  public JobStatus getJobStatus() {
```
Comment (ezimanyi) [inline orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/job/RunJobStageContext.java]: Any reason not to annotate the class with `@Getter` and `@Setter` instead of manually defining these?

Comment (emjburns) [reply orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/job/RunJobStageContext.java]: Do you know if this works `@JsonAnyGetter` for the `other()` properties?  That's why I didn't do it here

Comment (ezimanyi) [reply orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/job/RunJobStageContext.java]: I don't know for sure; my guess is that it would work because serialization/deserialization shouldn't care how the `getXXX` and `setXXX` methods were generated.  I think it just looks for methods with those names, then if absent would fall back to the `@JsonAny...` annotated methods.  But I also am not 100% sure of this so definitely could be wrong here...

Comment (emjburns) [reply orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/job/RunJobStageContext.java]: I'm going to leave this as it is for now, although I think you may be right.

**Comment by emjburns** [comment]
Got confirmation from a kubernetes run job user that `jobStatus.name` is present and is the correct name to cancel the job with.

**Comment by ethanfrogers** [comment]
LGTM!
